Arm backend: Add 16A8W FCNode support with BMM dependency fix

Ninja91 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #13802
* __->__ #13801
* #13800
* #13799
* #13798
* #13797
* #13796

Add 16A8W quantization support for FCNode operations with BMM dependency fix in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, and cat operations, extending int16 support to FCNode operations.

Changes:
- Add INT16 dtype validation support in op_bmm.py
- Add test_addmm_tensor_16a8w_tosa_INT test function
- Enable test_addmm.py in test targets configuration
- Fix BMM dependency for FCNode operations

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency.

Differential Revision: [D80512504](https://our.internmc.facebook.com/intern/diff/D80512504/)

cc @digantdesai @freddan80 @per @zingo @oscarandersson8218

Ninja91 committed:
Add 16A8W FCNode support with BMM dependency fix

Add 16A8W quantization support for FCNode operations with BMM dependency fix in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, and cat operations, extending int16 support to FCNode operations.

Changes:
- Add INT16 dtype validation support in op_bmm.py
- Add test_addmm_tensor_16a8w_tosa_INT test function
- Enable test_addmm.py in test targets configuration
- Fix BMM dependency for FCNode operations

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency.

Differential Revision: [D80512504](https://our.internmc.facebook.com/intern/diff/D80512504/)

[ghstack-poisoned]

digantdesai commented on backends/arm/operators/op_bmm.py:
```diff
@@ -93,7 +93,8 @@ def define_node(
             output_qparams = get_output_qparams(node)[0]
             final_output_scale = (
-                input_qparams[0].get_scale_per_tensor() * input_qparams[1].get_scale_per_tensor()  # type: ignore[possibly-undefined]  # pyre-ignore[61]
+                input_qparams[0].get_scale_per_tensor()
```
Fix mypy!

digantdesai reviewed (approved):
Please fix mypy before landing. Code dupe issue with the quant helper.

Ninja91 committed:
Update on "Arm backend: Add 16A8W FCNode support with BMM dependency fix"


Add 16A8W quantization support for FCNode operations with BMM dependency fix in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, and cat operations, extending int16 support to FCNode operations.

Changes:
- Add INT16 dtype validation support in op_bmm.py
- Add test_addmm_tensor_16a8w_tosa_INT test function
- Enable test_addmm.py in test targets configuration
- Fix BMM dependency for FCNode operations

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency.

Differential Revision: [D80512504](https://our.internmc.facebook.com/intern/diff/D80512504/)

cc digantdesai freddan80 per zingo oscarandersson8218

[ghstack-poisoned]

Ninja91 committed:
Update on "Arm backend: Add 16A8W FCNode support with BMM dependency fix"


Add 16A8W quantization support for FCNode operations with BMM dependency fix in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, and cat operations, extending int16 support to FCNode operations.

Changes:
- Add INT16 dtype validation support in op_bmm.py
- Add test_addmm_tensor_16a8w_tosa_INT test function
- Enable test_addmm.py in test targets configuration
- Fix BMM dependency for FCNode operations

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency.

Differential Revision: [D80512504](https://our.internmc.facebook.com/intern/diff/D80512504/)

cc digantdesai freddan80 per zingo oscarandersson8218

[ghstack-poisoned]

Ninja91 committed:
Update on "Arm backend: Add 16A8W FCNode support with BMM dependency fix"


Add 16A8W quantization support for FCNode operations with BMM dependency fix in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, and cat operations, extending int16 support to FCNode operations.

Changes:
- Add INT16 dtype validation support in op_bmm.py
- Add test_addmm_tensor_16a8w_tosa_INT test function
- Enable test_addmm.py in test targets configuration
- Fix BMM dependency for FCNode operations

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency.

Differential Revision: [D80512504](https://our.internmc.facebook.com/intern/diff/D80512504/)

cc digantdesai freddan80 per zingo oscarandersson8218

[ghstack-poisoned]

Ninja91 committed:
Update on "Arm backend: Add 16A8W FCNode support with BMM dependency fix"


Add 16A8W quantization support for FCNode operations with BMM dependency fix in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, and cat operations, extending int16 support to FCNode operations.

Changes:
- Add INT16 dtype validation support in op_bmm.py
- Add test_addmm_tensor_16a8w_tosa_INT test function
- Enable test_addmm.py in test targets configuration
- Fix BMM dependency for FCNode operations

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency.

Differential Revision: [D80512504](https://our.internmc.facebook.com/intern/diff/D80512504/)

cc digantdesai freddan80 per zingo oscarandersson8218

[ghstack-poisoned]

Ninja91 committed:
Update on "Arm backend: Add 16A8W FCNode support with BMM dependency fix"


Add 16A8W quantization support for FCNode operations with BMM dependency fix in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, and cat operations, extending int16 support to FCNode operations.

Changes:
- Add INT16 dtype validation support in op_bmm.py
- Add test_addmm_tensor_16a8w_tosa_INT test function
- Enable test_addmm.py in test targets configuration
- Fix BMM dependency for FCNode operations

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency.

Differential Revision: [D80512504](https://our.internmc.facebook.com/intern/diff/D80512504/)

cc digantdesai freddan80 per zingo oscarandersson8218

[ghstack-poisoned]

Ninja91 committed:
Update on "Arm backend: Add 16A8W FCNode support with BMM dependency fix"


Add 16A8W quantization support for FCNode operations with BMM dependency fix in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, and cat operations, extending int16 support to FCNode operations.

Changes:
- Add INT16 dtype validation support in op_bmm.py
- Add test_addmm_tensor_16a8w_tosa_INT test function
- Enable test_addmm.py in test targets configuration
- Fix BMM dependency for FCNode operations

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency.

Differential Revision: [D80512504](https://our.internmc.facebook.com/intern/diff/D80512504/)

cc digantdesai freddan80 per zingo oscarandersson8218

[ghstack-poisoned]

Ninja91 committed:
Update on "Arm backend: Add 16A8W FCNode support with BMM dependency fix"


Add 16A8W quantization support for FCNode operations with BMM dependency fix in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, and cat operations, extending int16 support to FCNode operations.

Changes:
- Add INT16 dtype validation support in op_bmm.py
- Add test_addmm_tensor_16a8w_tosa_INT test function
- Enable test_addmm.py in test targets configuration
- Fix BMM dependency for FCNode operations

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency.

Differential Revision: [D80512504](https://our.internmc.facebook.com/intern/diff/D80512504/)

cc digantdesai freddan80 per zingo oscarandersson8218

[ghstack-poisoned]

Ninja91 committed:
Update on "Arm backend: Add 16A8W FCNode support with BMM dependency fix"


Add 16A8W quantization support for FCNode operations with BMM dependency fix in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, and cat operations, extending int16 support to FCNode operations.

Changes:
- Add INT16 dtype validation support in op_bmm.py
- Add test_addmm_tensor_16a8w_tosa_INT test function
- Enable test_addmm.py in test targets configuration
- Fix BMM dependency for FCNode operations

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency.

Differential Revision: [D80512504](https://our.internmc.facebook.com/intern/diff/D80512504/)

cc digantdesai freddan80 per zingo oscarandersson8218

[ghstack-poisoned]

Ninja91 committed:
Update on "Arm backend: Add 16A8W FCNode support with BMM dependency fix"


Add 16A8W quantization support for FCNode operations with BMM dependency fix in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, and cat operations, extending int16 support to FCNode operations.

Changes:
- Add INT16 dtype validation support in op_bmm.py
- Add test_addmm_tensor_16a8w_tosa_INT test function
- Enable test_addmm.py in test targets configuration
- Fix BMM dependency for FCNode operations

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency.

Differential Revision: [D80512504](https://our.internmc.facebook.com/intern/diff/D80512504/)

cc digantdesai freddan80 per zingo oscarandersson8218

[ghstack-poisoned]
